bundle/permissions/mutator.go

```diff
@@ -0,0 +1,176 @@
+package permissions
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/databricks-sdk-go/service/workspace"
+)
+
+const CAN_MANAGE = "CAN_MANAGE"
+const CAN_VIEW = "CAN_VIEW"
+const CAN_RUN = "CAN_RUN"
+
+var allowedLevels = []string{CAN_MANAGE, CAN_VIEW, CAN_RUN}
+var levelsMap = map[string](map[string]string){
+	"jobs": {
+		CAN_MANAGE: "CAN_MANAGE",
+		CAN_VIEW:   "CAN_VIEW",
+		CAN_RUN:    "CAN_MANAGE_RUN",
+	},
+	"pipelines": {
+		CAN_MANAGE: "CAN_MANAGE",
+		CAN_VIEW:   "CAN_VIEW",
+		CAN_RUN:    "CAN_RUN",
+	},
+	"mlflow_experiments": {
+		CAN_MANAGE: "CAN_MANAGE",
+		CAN_VIEW:   "CAN_READ",
+	},
+	"mlflow_models": {
+		CAN_MANAGE: "CAN_MANAGE",
+		CAN_VIEW:   "CAN_READ",
+	},
+	"model_serving_endpoints": {
+		CAN_MANAGE: "CAN_MANAGE",
+		CAN_VIEW:   "CAN_VIEW",
+		CAN_RUN:    "CAN_QUERY",
+	},
+}
+
+type bundlePermissions struct{}
+
+func ApplyBundlePermissions() bundle.Mutator {
+	return &bundlePermissions{}
+}
+
+func (m *bundlePermissions) Apply(ctx context.Context, b *bundle.Bundle) error {
+	err := validate(b)
+	if err != nil {
+		return err
+	}
+
+	err = giveWriteAccessForWorkspaceRoot(ctx, b)
+	if err != nil {
+		return err
+	}
+
+	applyForJobs(ctx, b)
+	applyForPipelines(ctx, b)
+	applyForMlModels(ctx, b)
+	applyForMlExperiments(ctx, b)
+	applyForModelServiceEndpoints(ctx, b)
+
+	return nil
+}
+
+func validate(b *bundle.Bundle) error {
+	for _, p := range b.Config.Permissions {
+		if !slices.Contains(allowedLevels, p.Level) {
+			return fmt.Errorf("invalid permission level: %s, allowed values: [%s]", p.Level, strings.Join(allowedLevels, ", "))
+		}
+	}
+
+	return nil
+}
+
+func giveWriteAccessForWorkspaceRoot(ctx context.Context, b *bundle.Bundle) error {
+	permissions := make([]workspace.WorkspaceObjectAccessControlRequest, 0)
+
+	for _, p := range b.Config.Permissions {
+		if p.Level != CAN_MANAGE {
+			continue
+		}
+
+		permissions = append(permissions, workspace.WorkspaceObjectAccessControlRequest{
+			GroupName:            p.GroupName,
+			UserName:             p.UserName,
+			ServicePrincipalName: p.ServicePrincipalName,
+			PermissionLevel:      "CAN_MANAGE",
+		})
+	}
+
+	if len(permissions) == 0 {
+		return nil
+	}
+
+	w := b.WorkspaceClient().Workspace
+	obj, err := w.GetByPath(ctx, b.Config.Workspace.RootPath)
```

Comment:
Has the root path already been created for new bundle deployments at this point?

Reply:
It might yet not been, so moved it to after files.Upload in deploy phase